usr/src/lib/libzfs/common/libzfs_dataset.c

```diff
@@ -3642,17 +3642,13 @@ zfs_promote(zfs_handle_t *zhp)
 		return (zfs_error(hdl, EZFS_BADTYPE, errbuf));
 	}
 
-	(void) strlcpy(parent, zhp->zfs_dmustats.dds_origin, sizeof (parent));
-	if (parent[0] == '\0') {
+	if (zhp->zfs_dmustats.dds_origin[0] == '\0') {
 		zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
 		    "not a cloned filesystem"));
 		return (zfs_error(hdl, EZFS_BADTYPE, errbuf));
 	}
 
-	(void) strlcpy(zc.zc_value, zhp->zfs_dmustats.dds_origin,
-	    sizeof (zc.zc_value));
-	(void) strlcpy(zc.zc_name, zhp->zfs_name, sizeof (zc.zc_name));
-	ret = zfs_ioctl(hdl, ZFS_IOC_PROMOTE, &zc);
+	ret = lzc_promote(zhp->zfs_name, snapname, sizeof(snapname));
 
 	if (ret != 0) {
 		int save_errno = errno;
```

Comment:
I was hoping nobody would pick up on this.

So here's the deal: most of the functions in libzfs_core leave errno set to whatever value was set by iocl(2).  Even though they return the error many calls in libzfs rely on this being set.

What I'd rather is to address that in a different changeset, since there are a number of areas in the code that need to be updated, and I feel that's outside the scope of this change.

While it's not exactly right, it's not exactly wrong either.